File: leetcode-141.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def hasCycle(self, head: ListNode) -> bool:
        dic = dict()
        def dfs(u):
            if u == None:
                return False
            if u in dic:
                return True
            dic[u] = 1
            if not dfs(u.next):
                return True
            return False
        return dfs(head)

File: leetcode/leetcode-141.cpp
/*
 * @File Name          :leetcode-141.cpp
 * @Author             :cc
 * @Version            :1.0.0
 * @Date               :2022/1/7 9:22
 * @Description        :
 * @Function List      :
 * @History            :
 */

#include <cstddef>


struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};


class Solution {
public:
    bool hasCycle(ListNode *head) {
        if(head == nullptr) return false;
        ListNode* slow = head, *fast = head->next;
        while (slow != nullptr && fast!= nullptr&& slow != fast){
            slow = slow->next;
            if(fast->next != nullptr)
                fast = fast->next->next;
            else fast = nullptr;
        }return slow == fast;
    }
};